allennlp/modules/conditional_random_field.py

```diff
@@ -0,0 +1,170 @@
+"""
+Conditional random field
+"""
+
+import torch
+
+def log_sum_exp(x: torch.Tensor) -> torch.Tensor:  # pylint: disable=invalid-name
+    """
+    Computes log(sum(exp(x))) in a numerically stable way,
+    where the sum is only over the last dimension.
+    """
+    maxes, _ = torch.max(x, -1, keepdim=True)
+    exps = torch.exp(x - maxes)
+
+    return maxes.squeeze(-1) + torch.log(torch.sum(exps, -1))
+
+
+class ConditionalRandomField(torch.nn.Module):
+    """
+    This module uses the "forward-backward" algorithm to compute
+    the log-likelihood of its inputs assuming a conditional random field model.
+
+    See, e.g. http://www.cs.columbia.edu/~mcollins/fb.pdf
+
+    Parameters
+    ----------
+    num_tags: int, required
+        The number of tags.
+    start_tag: int, required
+        The id of the sentinel <START> tag.
+    stop_tag: int, required
+        The id of the sentinel <STOP> tag.
+    """
+    def __init__(self,
+                 num_tags: int,
+                 start_tag: int,
+                 stop_tag: int) -> None:
+        super().__init__()
+
+        self.num_tags = num_tags
+        self.start_tag = start_tag
+        self.stop_tag = stop_tag
+
+        # transitions[i, j] is the logit for transitioning to state i from state j
+        self.transitions = torch.nn.Parameter(torch.randn(num_tags, num_tags))
+
+        # We never transition to the start tag and we never transition from the stop tag.
+        self.transitions.data[start_tag, :] = -10000
```

Comment:
Hmm, I don't think this will fix the value to `-10000`, because the parameter is still trainable - it probably does what you want, i.e massively biasing the model to learn not to transition to these tags, but I could imagine it makes learning a bit unstable because of the large values. Given you calculate the potentials for these separately below anyway, it might be better to just remove them from the parameter and calculate it with a fixed, constant value.

Reply:
but this definitely does affect my regularization, which is bad. let me think about that